crossbar/router/test/test_broker.py

```diff
@@ -511,6 +512,197 @@ class TestSession(ApplicationSession):
             self.assertTrue(events[3].correlation_is_last)
 
+    def test_subscribe_unsubscribe(self):
+        """
+        Make sure the wamp.* event sequence delivered as part of a subscribe/unsubscribe sequence
+        is correct and contains the correct session id and subscription id values.
+        """
+        router = self.router
+        test = self
+
+        class Session(mock.MagicMock):
+
+            _private = []
+            _events = []
+            _authrole = 'trusted'
+            _session_id = 0
+
+            def __init__(self, *args, **kwargs):
+                super().__init__(*args, **kwargs)
+
+            def send(self, *args, **argv):
+                self._private.append(args[0])
+
+            def publish(self, *args, **argv):
+                self._events.append([args, argv])
+
+        class TestSession(ApplicationSession):
+
+            def __init__(self, *args, **kw):
+                super().__init__(*args, **kw)
+                self._service_session = Session()
+                self._router = router
+
+            def onJoin(self, details):
+                self._service_session._session_id = details.session
+                router.attach(self._service_session)
+
+                router._broker._router._realm.session = self._service_session
+                subscription = message.Subscribe(self._service_session._session_id, u'com.example.test1')
+                router._broker.processSubscribe(self._service_session, subscription)
+                subscription = message.Subscribe(self._service_session._session_id, u'com.example.test2')
+                router._broker.processSubscribe(self._service_session, subscription)
+                subscription = message.Subscribe(self._service_session._session_id, u'com.example.test3')
+                router._broker.processSubscribe(self._service_session, subscription)
+
+                subscriptions = []
+                for obj in list(self._service_session._private):
+                    subscription = message.Unsubscribe(self._service_session._session_id, subscription=obj.subscription)
+                    router._broker.processUnsubscribe(self._service_session, subscription)
+                    subscriptions.append(obj.subscription)
+
+                def all_done():
+
+                    created = list(subscriptions)
+                    subscribes = list(subscriptions)
+                    unsubscribes = list(subscriptions)
+                    deletes = list(subscriptions)
+
+                    for args, argv in self._service_session._events:
+                        if args[0] == 'wamp.subscription.on_create':
+                            test.assertEqual(args[1], self._service_session._session_id, 'on_create: session id is incorrect!')
+                            test.assertTrue(args[2]['id'] in created, 'on_create: subscription id is incorrect!')
+                            created.remove(args[2]['id'])
+
```

Comment:
This sort of pattern always makes me wonder: should this be like 4 separate tests instead? (Maybe not, but ..)

Reply:
Okay, fair enough.